WikiWatershed/model-my-watershed
# Add support for creating AoI via uploaded Shapefile and GeoJSON

## Overview

Adds support for creating AoIs from uploaded shapefiles and geojson.

The browser DataTransfer and FileReader APIs are used to upload and read
in the file.

At the moment, only the first feature from the uploaded file is used.
This may change in the future.

The UI for this feature is temporary. It will be moved once #1743
is implemented

Tested in Chrome, Safari, Firefox, and IE11.

Connects #1739

### Demo

![mmw1](https://cloud.githubusercontent.com/assets/1042475/24252848/06142fb8-0fb5-11e7-88fc-d22bcbaaa880.gif)

### Notes

There are a couple follow up issues:

- Move the UI to the sidebar and add a file selector input (per the wireframes).
- Add more validation (check that coordinates are geographic, check that the shape is in the U.S., implement a better file type check than looking at the extension). It may be a good idea to create a Backbone model for uploaded data and move the validation there.

## Testing Instructions

I didn't provide any test data because I thought it would be better to test it with more datasets than the ones I created.

Try to upload the following geometries (either GeoJSON or a shapefile), and check that you get the expected results:

- A small polygon; should proceed to analyze
- A very large polygon; should add the shape to the map but display an error reporting that it can't be analyzed because it's too large
- A file that isn't a shapefile or geojson; should get an error that the file type is not supported.
- A file that is larger than 200 MB; should get an error that the file is too large.
- A self-intersecting geometry; should add the shape to the map but display an error reporting that it can't be analyzed because it's self-intersecting.
- A file with multiple features; the first one should be added to the map.
